### PR TITLE
✨ Set & apply routing table rules

### PR DESCRIPTION
**Pull request by ma1ted**
See #22.

Uses rtnetlink to add a TUN interface with a specified IP to the routing table. Currently routes all traffic through the interface.

### CODE REVIEW

File: burrow/Cargo.toml [hunk 1]
```diff
@@ -7,5 +7,7 @@ edition = "2021"
 crate-type = ["lib", "staticlib"]
 
 [dependencies]
+ipnetwork = "0.20.0"
+rtnetlink = "0.12.0"
```
Comment (Muirrum) [inline burrow/Cargo.toml:11]: Curious why `burrow` needs rtnetlink and ipnetwork?

File: tun/src/unix/linux/mod.rs [hunk 2]
```diff
@@ -68,6 +67,17 @@ impl TunInterface {
...
+    pub fn set_iface_up(&self) {
+        std::process::Command::new("ip")
+            .arg("link")
+            .arg("set")
+            .arg("dev")
+            .arg(&self.name().unwrap())
+            .arg("up")
+            .output()?;
+    }
+
```
Comment (Muirrum) [inline tun/src/unix/linux/mod.rs:80]: We should not be calling other commands directly.

**Review by Muirrum — CHANGES_REQUESTED** [review]

File: tun/src/unix/linux/mod.rs (hunk 2 above)
Comment (ma1ted) [reply tun/src/unix/linux/mod.rs:80]: I'm aware it's not ideal.
I lost 3 hours to this reading linux header files; `/dev/net/tun` just doesen't want to listen to me when I set flags. It's quite a travesty actually.

File: burrow/Cargo.toml (hunk 1 above)
Comment (ma1ted) [reply burrow/Cargo.toml:11]: Forgot to remove those; thanks!

Comment (Muirrum) [reply burrow/Cargo.toml:11]: https://docs.rs/rtnetlink/latest/rtnetlink/struct.LinkSetRequest.html#method.up

Look at using this method instead

File: tun/src/unix/linux/mod.rs (hunk 2 above)
Comment (Muirrum) [reply tun/src/unix/linux/mod.rs:80]: https://docs.rs/rtnetlink/latest/rtnetlink/struct.LinkSetRequest.html#method.up

Look at using this method instead

Comment (ma1ted) [reply tun/src/unix/linux/mod.rs:80]: It's definitely only a temporary solution.

File: tun/src/unix/linux/mod.rs [hunk 3]
```diff
@@ -129,7 +118,8 @@ impl TunInterface {
         let (connection, handle, _) = new_connection().expect("a new netlink connection");
         let netlink_connection_handle = tokio::task::spawn(connection);
 
-        self.set_iface_up().expect("an active interface");
+        let index = self.index().expect("a valid interface index");
+        _ = handle.link().set(index as u32).up().execute().await;
```
Comment (Muirrum) [inline tun/src/unix/linux/mod.rs:122]: I think you were on the right track to separate this out into a different function -- we probably want to control the state of the link independently of the routing tables. What if we're modifying routing before we want the interface up?

**Review by Muirrum — CHANGES_REQUESTED** [review]

**Comment by ma1ted** [comment]
I'll add these changes in a new PR

**Pull request closed without merging**